cryptosuite/bbsplussignaturesuite.go

```diff
@@ -0,0 +1,355 @@
+package cryptosuite
+
+import (
+	gocrypto "crypto"
+	"encoding/base64"
+
+	"github.com/TBD54566975/ssi-sdk/crypto"
+	. "github.com/TBD54566975/ssi-sdk/util"
+	"github.com/goccy/go-json"
+	"github.com/pkg/errors"
+)
+
+const (
+	BBSSecurityContext                             string        = "https://w3id.org/security/bbs/v1"
+	BBSPlusSignature2020                           SignatureType = "BbsBlsSignature2020"
+	BBSPlusSignatureSuiteID                        string        = "https://w3c-ccg.github.io/ldp-bbs2020/#the-bbs-signature-suite-2020"
+	BBSPlusSignatureSuiteType                      LDKeyType     = BLS12381G2Key2020
+	BBSPlusSignatureSuiteCanonicalizationAlgorithm string        = "https://w3id.org/security#URDNA2015"
+	// BBSPlusSignatureSuiteDigestAlgorithm uses https://www.rfc-editor.org/rfc/rfc4634
+	BBSPlusSignatureSuiteDigestAlgorithm gocrypto.Hash = gocrypto.BLAKE2b_384
+	// BBSPlusSignatureSuiteProofAlgorithm  uses https://www.rfc-editor.org/rfc/rfc7797
+)
+
+type BBSPlusSignatureSuite struct {
+	CryptoSuiteProofType
+}
+
+func GetBBSPlusSignatureSuite() CryptoSuite {
+	return new(BBSPlusSignatureSuite)
+}
+
+// CryptoSuiteInfo interface
+
+func (BBSPlusSignatureSuite) ID() string {
+	return BBSPlusSignatureSuiteID
+}
+
+func (BBSPlusSignatureSuite) Type() LDKeyType {
+	return BBSPlusSignatureSuiteType
+}
+
+func (BBSPlusSignatureSuite) CanonicalizationAlgorithm() string {
+	return BBSPlusSignatureSuiteCanonicalizationAlgorithm
+}
+
+func (BBSPlusSignatureSuite) MessageDigestAlgorithm() gocrypto.Hash {
+	return BBSPlusSignatureSuiteDigestAlgorithm
+}
+
+func (BBSPlusSignatureSuite) SignatureAlgorithm() SignatureType {
+	return BBSPlusSignature2020
+}
+
+func (BBSPlusSignatureSuite) RequiredContexts() []string {
+	return []string{BBSSecurityContext}
+}
+
+func (b BBSPlusSignatureSuite) Sign(s Signer, p Provable) error {
+	// create proof before CVH
+	// TODO(gabe) support required reveal values
+	proof := b.createProof(s.GetKeyID(), s.GetProofPurpose(), nil)
+
+	// prepare proof options
+	contexts, err := GetContextsFromProvable(p)
+	if err != nil {
+		return errors.Wrap(err, "could not get contexts from provable")
+	}
+
+	// make sure the suite's context(s) are included
+	contexts = ensureRequiredContexts(contexts, b.RequiredContexts())
+	opts := &ProofOptions{Contexts: contexts}
+
+	// 3. tbs value as a result of cvh
+	tbs, err := b.CreateVerifyHash(p, proof, opts)
+	if err != nil {
+		return errors.Wrap(err, "create verify hash algorithm failed")
+	}
+
+	// 4 & 5. create the signature over the provable data as a JWS
```

Comment:
I don't follow why this is as a JWS... isn't this only the digital signature (i.e. the result of hashing, and then signing the data)?

Reply:
sorry, this comment is wrong it's not a jws. will update.